### PR TITLE
Improve performance of node selector matcher

### LINKED ISSUE #107504: Improve performance of node selector matcher
### What would you like to be added?

Cache the actual operation of a requirement (for example, as a function) instead of storing it as a string

https://github.com/kubernetes/kubernetes/blob/bf8c918e0be340637e63520eb3b58df493a230f9/vendor/sigs.k8s.io/kustomize/kyaml/yaml/internal/k8sgen/pkg/labels/selector.go#L145:6

### Why is this needed?

When we parse a node selector, we transform the API operation string into another internal string.

Even though we are saving some computing time because the matcher doesn't do validations anymore, we still need to re-parse the operation.

Parsing and matching node selectors is a very common operation throughout the code base, but in particular in scheduler.
We could run the scheduling benchmarks to determine if the change is worth:
- https://github.com/kubernetes/kubernetes/tree/master/test/integration/scheduler_perf
- https://github.com/kubernetes/kubernetes/blob/bf8c918e0be340637e63520eb3b58df493a230f9/pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go#L1062

### PR DESCRIPTION
**Pull request by yuzhiquan**
#### What type of PR is this?

/kind cleanup
#### Which issue(s) this PR fixes:

Fixes #107504 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```
/sig api-machinery
/sig scheduling
/assign @alculquicondor

### CODE REVIEW

**Comment by yuzhiquan** [comment]
scheduling benchmarks on my laptop

before
```
/private/var/folders/jj/hmhqkjy11f7d0jltb6mpr8c00000gn/T/GoLand/___BenchmarkFilter_in_k8s_io_kubernetes_pkg_scheduler_framework_plugins_podtopologyspread.test -test.v -test.paniconexit0 -test.bench ^\QBenchmarkFilter\E$ -test.run ^$
goos: darwin
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread
cpu: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
BenchmarkFilter
BenchmarkFilter/1000nodes/single-constraint-zone
BenchmarkFilter/1000nodes/single-constraint-zone-8         	    1759	    750968 ns/op
BenchmarkFilter/1000nodes/single-constraint-zone/Clone
BenchmarkFilter/1000nodes/single-constraint-zone/Clone-8   	  752032	      1662 ns/op
BenchmarkFilter/1000nodes/single-constraint-node
BenchmarkFilter/1000nodes/single-constraint-node-8         	    1263	    953243 ns/op
BenchmarkFilter/1000nodes/single-constraint-node/Clone
BenchmarkFilter/1000nodes/single-constraint-node/Clone-8   	   12343	     98375 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node
BenchmarkFilter/1000nodes/two-Constraints-zone-node-8      	     921	   1236147 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-8         	   12152	     96975 ns/op
PASS
```
after
```
/private/var/folders/jj/hmhqkjy11f7d0jltb6mpr8c00000gn/T/GoLand/___BenchmarkFilter_in_k8s_io_kubernetes_pkg_scheduler_framework_plugins_podtopologyspread.test -test.v -test.paniconexit0 -test.bench ^\QBenchmarkFilter\E$ -test.run ^$
goos: darwin
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread
cpu: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
BenchmarkFilter
BenchmarkFilter/1000nodes/single-constraint-zone
BenchmarkFilter/1000nodes/single-constraint-zone-8         	    1766	    680068 ns/op
BenchmarkFilter/1000nodes/single-constraint-zone/Clone
BenchmarkFilter/1000nodes/single-constraint-zone/Clone-8   	  772818	      1655 ns/op
BenchmarkFilter/1000nodes/single-constraint-node
BenchmarkFilter/1000nodes/single-constraint-node-8         	    1450	    741098 ns/op
BenchmarkFilter/1000nodes/single-constraint-node/Clone
BenchmarkFilter/1000nodes/single-constraint-node/Clone-8   	   12687	     95398 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node
BenchmarkFilter/1000nodes/two-Constraints-zone-node-8      	     979	   1213292 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-8         	   12525	     95070 ns/op
PASS
```

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers.go [hunk 1]
```diff
@@ -47,20 +47,11 @@ func LabelSelectorAsSelector(ps *LabelSelector) (labels.Selector, error) {
...
-			op = selection.Exists
-		case LabelSelectorOpDoesNotExist:
-			op = selection.DoesNotExist
-		default:
+		err, op := selection.TransToOperator(string(expr.Operator))
+		if err != nil {
 			return nil, fmt.Errorf("%q is not a valid pod selector operator", expr.Operator)
 		}
-		r, err := labels.NewRequirement(expr.Key, op, append([]string(nil), expr.Values...))
+		r, err := labels.NewRequirement(expr.Key, *op, append([]string(nil), expr.Values...))
```
Comment (denkensk) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers.go:54]: `NewRequirement` also uses `switch` to justify the type of `operation` after  `transToOperator` already knows the type. I guess if `NewRequirement` also has some room for optimization ？

**Review by MadhavJivrajani — COMMENTED** [review]
A few more considerations:

Since benchmarks can be tricky to gauge, I benchmarked over 10 and 50 runs and used [`benchstat`](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) to compare the performance before this change and after this change. The command I used to run the benchmark is:
```
go test -run XXX -bench ^BenchmarkFilter$ -count 10 >> ~/new_sched_10.txt
```
and similarly for the benchmark before this change. On comparing the results, we get the following

For 50 runs:
```
benchstat -sort delta ~/old_sched.txt ~/new_sched.txt
name                                                old time/op  new time/op   delta
Filter/1000nodes/single-constraint-node/Clone-8     95.0µs ± 1%  100.4µs ±15%   +5.70%  (p=0.000 n=41+47)
Filter/1000nodes/two-Constraints-zone-node-8        1.38ms ± 8%   1.43ms ±23%   +3.95%  (p=0.001 n=46+44)
Filter/1000nodes/two-Constraints-zone-node/Clone-8  96.8µs ± 2%   98.1µs ± 2%   +1.37%  (p=0.000 n=36+41)
Filter/1000nodes/single-constraint-zone/Clone-8     1.66µs ± 7%   1.62µs ±11%     ~     (p=0.964 n=30+50)
Filter/1000nodes/single-constraint-node-8           1.14ms ±17%   1.24ms ±48%     ~     (p=0.198 n=48+50)
Filter/1000nodes/single-constraint-zone-8           1.42ms ±26%   0.97ms ± 2%  -32.11%  (p=0.000 n=49+46)
```
For 10 runs:
```
benchstat -sort delta ~/old_sched_10.txt ~/new_sched_10.txt
name                                                old time/op  new time/op  delta
Filter/1000nodes/single-constraint-zone-8           1.06ms ± 5%  1.19ms ±11%  +12.27%  (p=0.000 n=9+10)
Filter/1000nodes/two-Constraints-zone-node-8        1.39ms ± 2%  1.48ms ±16%   +6.07%  (p=0.019 n=10+10)
Filter/1000nodes/single-constraint-zone/Clone-8     1.73µs ±17%  1.54µs ± 4%     ~     (p=0.052 n=10+9)
Filter/1000nodes/single-constraint-node-8           1.19ms ± 5%  1.30ms ±13%     ~     (p=0.052 n=10+10)
Filter/1000nodes/single-constraint-node/Clone-8     88.8µs ±24%  93.1µs ±11%     ~     (p=0.095 n=9+10)
Filter/1000nodes/two-Constraints-zone-node/Clone-8  98.3µs ± 2%  97.9µs ± 3%     ~     (p=0.645 n=8+8)
```
Additionally, we also have more allocations happening after the change is made (my guess is because of the global map of operators acting as a cache?)
Old (this is just for one run of the benchmark to get the allocs):
```
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread
cpu: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
BenchmarkFilter/1000nodes/single-constraint-zone-8         	    1237	   1015809 ns/op	  413811 B/op	    2065 allocs/op
BenchmarkFilter/1000nodes/single-constraint-zone/Clone-8   	  492946	      2353 ns/op	    1495 B/op	       9 allocs/op
BenchmarkFilter/1000nodes/single-constraint-node-8         	    1064	   1096974 ns/op	  511108 B/op	    2065 allocs/op
BenchmarkFilter/1000nodes/single-constraint-node/Clone-8   	    8775	    133160 ns/op	   99112 B/op	      10 allocs/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node-8      	     885	   1367161 ns/op	  673509 B/op	    4576 allocs/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-8         	    9243	    112293 ns/op	   99160 B/op	      11 allocs/op
```
New:
```
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread
cpu: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
BenchmarkFilter/1000nodes/single-constraint-zone-8         	    1257	    972454 ns/op	  413979 B/op	    2067 allocs/op
BenchmarkFilter/1000nodes/single-constraint-zone/Clone-8   	  510312	      1960 ns/op	    1495 B/op	       9 allocs/op
BenchmarkFilter/1000nodes/single-constraint-node-8         	    1135	   1218474 ns/op	  511116 B/op	    2067 allocs/op
BenchmarkFilter/1000nodes/single-constraint-node/Clone-8   	   12579	     95917 ns/op	   99112 B/op	      10 allocs/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node-8      	     854	   1356366 ns/op	  672331 B/op	    4580 allocs/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-8         	   12871	     90156 ns/op	   99160 B/op	      11 allocs/op
```
Currently, I'm not sure if the speedups are significant enough to warrant additional allocs (which again aren't all that much). Very possible I'm missing something here, please lmk if that's the case. 

It also might be worth exploring what the results are by increasing the benchtime.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers.go (hunk 1 above)
Comment (yuzhiquan) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers.go:54]: The switch in `NewRequirement ` is likely to verify  values in every Operator, not trans other string to Operator.Did i miss something?

**Comment by yuzhiquan** [comment]
> A few more considerations:
> 
> Since benchmarks can be tricky to gauge, I benchmarked over 10 and 50 runs and used [`benchstat`](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) to compare the performance before this change and after this change. The command I used to run the benchmark is:
> 
> ```
> go test -run XXX -bench ^BenchmarkFilter$ -count 10 >> ~/new_sched_10.txt
> ```
> 
> and similarly for the benchmark before this change. On comparing the results, we get the following
> 
> For 50 runs:
> 
> ```
> benchstat -sort delta ~/old_sched.txt ~/new_sched.txt
> name                                                old time/op  new time/op   delta
> Filter/1000nodes/single-constraint-node/Clone-8     95.0µs ± 1%  100.4µs ±15%   +5.70%  (p=0.000 n=41+47)
> Filter/1000nodes/two-Constraints-zone-node-8        1.38ms ± 8%   1.43ms ±23%   +3.95%  (p=0.001 n=46+44)
> Filter/1000nodes/two-Constraints-zone-node/Clone-8  96.8µs ± 2%   98.1µs ± 2%   +1.37%  (p=0.000 n=36+41)
> Filter/1000nodes/single-constraint-zone/Clone-8     1.66µs ± 7%   1.62µs ±11%     ~     (p=0.964 n=30+50)
> Filter/1000nodes/single-constraint-node-8           1.14ms ±17%   1.24ms ±48%     ~     (p=0.198 n=48+50)
> Filter/1000nodes/single-constraint-zone-8           1.42ms ±26%   0.97ms ± 2%  -32.11%  (p=0.000 n=49+46)
> ```
> 
> For 10 runs:
> 
> ```
> benchstat -sort delta ~/old_sched_10.txt ~/new_sched_10.txt
> name                                                old time/op  new time/op  delta
> Filter/1000nodes/single-constraint-zone-8           1.06ms ± 5%  1.19ms ±11%  +12.27%  (p=0.000 n=9+10)
> Filter/1000nodes/two-Constraints-zone-node-8        1.39ms ± 2%  1.48ms ±16%   +6.07%  (p=0.019 n=10+10)
> Filter/1000nodes/single-constraint-zone/Clone-8     1.73µs ±17%  1.54µs ± 4%     ~     (p=0.052 n=10+9)
> Filter/1000nodes/single-constraint-node-8           1.19ms ± 5%  1.30ms ±13%     ~     (p=0.052 n=10+10)
> Filter/1000nodes/single-constraint-node/Clone-8     88.8µs ±24%  93.1µs ±11%     ~     (p=0.095 n=9+10)
> Filter/1000nodes/two-Constraints-zone-node/Clone-8  98.3µs ± 2%  97.9µs ± 3%     ~     (p=0.645 n=8+8)
> ```
> 
> Additionally, we also have more allocations happening after the change is made (my guess is because of the global map of operators acting as a cache?) Old (this is just for one run of the benchmark to get the allocs):
> 
> ```
> goos: linux
> goarch: amd64
> pkg: k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread
> cpu: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
> BenchmarkFilter/1000nodes/single-constraint-zone-8         	    1237	   1015809 ns/op	  413811 B/op	    2065 allocs/op
> BenchmarkFilter/1000nodes/single-constraint-zone/Clone-8   	  492946	      2353 ns/op	    1495 B/op	       9 allocs/op
> BenchmarkFilter/1000nodes/single-constraint-node-8         	    1064	   1096974 ns/op	  511108 B/op	    2065 allocs/op
> BenchmarkFilter/1000nodes/single-constraint-node/Clone-8   	    8775	    133160 ns/op	   99112 B/op	      10 allocs/op
> BenchmarkFilter/1000nodes/two-Constraints-zone-node-8      	     885	   1367161 ns/op	  673509 B/op	    4576 allocs/op
> BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-8         	    9243	    112293 ns/op	   99160 B/op	      11 allocs/op
> ```
> 
> New:
> 
> ```
> goos: linux
> goarch: amd64
> pkg: k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread
> cpu: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
> BenchmarkFilter/1000nodes/single-constraint-zone-8         	    1257	    972454 ns/op	  413979 B/op	    2067 allocs/op
> BenchmarkFilter/1000nodes/single-constraint-zone/Clone-8   	  510312	      1960 ns/op	    1495 B/op	       9 allocs/op
> BenchmarkFilter/1000nodes/single-constraint-node-8         	    1135	   1218474 ns/op	  511116 B/op	    2067 allocs/op
> BenchmarkFilter/1000nodes/single-constraint-node/Clone-8   	   12579	     95917 ns/op	   99112 B/op	      10 allocs/op
> BenchmarkFilter/1000nodes/two-Constraints-zone-node-8      	     854	   1356366 ns/op	  672331 B/op	    4580 allocs/op
> BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-8         	   12871	     90156 ns/op	   99160 B/op	      11 allocs/op
> ```
> 
> Currently, I'm not sure if the speedups are significant enough to warrant additional allocs (which again aren't all that much). Very possible I'm missing something here, please lmk if that's the case.
> 
> It also might be worth exploring what the results are by increasing the benchtime.

That's impressive!thanks for the test, i will give another try!

**Comment by yuzhiquan** [comment]
> Replacing switch for maps is not very likely to give meaningful speedups.
> 
> Also, what I want to avoid is the switch during `Requirement.Matches` as that's the one that we want to use repeatedly.
> 
> Not sure how to explain my idea better: when creating a `Requirement` (in `NewRequirement`), we know what the operation is, so we can store a matcher `func` in the `Requirement` that we can directly call, instead of storing the operation `string` and doing the switch again.
> 
> An alternative approach is that `Requirement` is an interface and there are different implementations of it, but I think this is overkill.

I see, i will give another try soon!

**Comment by yuzhiquan** [comment]
> benchstat

in this commit, we get this benchmark:
```shell
> /Users/yuzhiquan/go/bin/benchstat old_50.txt new_50.txt
name                                                old time/op  new time/op  delta
Filter/1000nodes/single-constraint-zone-8            847µs ±44%   774µs ±78%  -8.58%  (p=0.001 n=49+45)
Filter/1000nodes/single-constraint-zone/Clone-8     1.62µs ± 5%  1.61µs ± 9%    ~     (p=0.124 n=45+50)
Filter/1000nodes/single-constraint-node-8            878µs ±22%   886µs ±39%    ~     (p=0.638 n=50+50)
Filter/1000nodes/single-constraint-node/Clone-8     90.3µs ± 5%  89.2µs ± 5%    ~     (p=0.085 n=49+43)
Filter/1000nodes/two-Constraints-zone-node-8        1.15ms ±16%  1.14ms ±16%    ~     (p=0.159 n=48+48)
Filter/1000nodes/two-Constraints-zone-node/Clone-8  89.6µs ± 8%  91.8µs ± 9%  +2.48%  (p=0.001 n=46+45)
```

File: staging/src/k8s.io/apimachinery/pkg/selection/operator.go [hunk 2]
```diff
@@ -31,3 +36,25 @@ const (
 	GreaterThan  Operator = "gt"
 	LessThan     Operator = "lt"
 )
+
+var (
+	OperatorMap map[string]Operator = map[string]Operator{
```
Comment (alculquicondor) [inline staging/src/k8s.io/apimachinery/pkg/selection/operator.go:41]: Prefer to leave this file untouched.

We don't need to replace every switch by maps. That's not necessarily going to give a performance boost (although you could try and show results from experiments).

File: staging/src/k8s.io/apimachinery/pkg/labels/selector.go [hunk 3]
```diff
@@ -40,11 +40,25 @@ var (
 		string(selection.GreaterThan), string(selection.LessThan),
 	}
 	validRequirementOperators = append(binaryOperators, unaryOperators...)
+
+	OperatorMapMatch = map[selection.Operator]MatchFunc{
```
Comment (alculquicondor) [inline staging/src/k8s.io/apimachinery/pkg/labels/selector.go:44]: This doesn't need to be exported

File: staging/src/k8s.io/apimachinery/pkg/labels/selector.go [hunk 4]
```diff
@@ -954,3 +935,52 @@ func SelectorFromValidatedSet(ls Set) Selector {
 func ParseToRequirements(selector string, opts ...field.PathOption) ([]Requirement, error) {
 	return parse(selector, field.ToPath(opts...))
 }
+
+func InEqualsDoubleEquals(r *Requirement, ls Labels) bool {
```
Comment (alculquicondor) [inline staging/src/k8s.io/apimachinery/pkg/labels/selector.go:939]: all these functions don't need to be exported

File: staging/src/k8s.io/apimachinery/pkg/labels/selector.go [hunk 5]
```diff
@@ -192,7 +207,12 @@ func NewRequirement(key string, op selection.Operator, vals []string, opts ...fi
 			allErrs = append(allErrs, err)
 		}
 	}
-	return &Requirement{key: key, operator: op, strValues: vals}, allErrs.ToAggregate()
+
+	var matchF MatchFunc
+	if o, ok := OperatorMapMatch[op]; ok {
+		matchF = o
+	}
```
Comment (alculquicondor) [inline staging/src/k8s.io/apimachinery/pkg/labels/selector.go:214]:
```suggestion
	matchF := OperatorMapMatch[op]
```

**Comment by yuzhiquan** [comment]
New benchmark results:
```shell
> /Users/yuzhiquan/go/bin/benchstat old_50.txt new_50-2.txt
name                                                old time/op  new time/op  delta
Filter/1000nodes/single-constraint-zone-8            847µs ±44%   773µs ±32%    ~     (p=0.052 n=49+44)
Filter/1000nodes/single-constraint-zone/Clone-8     1.62µs ± 5%  1.69µs ±14%  +4.26%  (p=0.000 n=45+48)
Filter/1000nodes/single-constraint-node-8            878µs ±22%   924µs ±29%    ~     (p=0.076 n=50+49)
Filter/1000nodes/single-constraint-node/Clone-8     90.3µs ± 5%  91.8µs ±10%  +1.65%  (p=0.031 n=49+46)
Filter/1000nodes/two-Constraints-zone-node-8        1.15ms ±16%  1.19ms ±11%  +3.28%  (p=0.002 n=48+42)
Filter/1000nodes/two-Constraints-zone-node/Clone-8  89.6µs ± 8%  89.6µs ± 7%    ~     (p=0.215 n=46+43)
```

**Comment by yuzhiquan** [comment]
New benchmark result:
```shell
> /Users/yuzhiquan/go/bin/benchstat old_50-2.txt new_50-2.txt
name                                                old time/op  new time/op  delta
Filter/1000nodes/single-constraint-zone-8            793µs ±36%   773µs ±32%    ~     (p=0.777 n=50+44)
Filter/1000nodes/single-constraint-zone/Clone-8     1.58µs ± 4%  1.69µs ±14%  +6.56%  (p=0.000 n=46+48)
Filter/1000nodes/single-constraint-node-8            897µs ±30%   924µs ±29%    ~     (p=0.289 n=47+49)
Filter/1000nodes/single-constraint-node/Clone-8     95.5µs ±20%  91.8µs ±10%  -3.84%  (p=0.007 n=47+46)
Filter/1000nodes/two-Constraints-zone-node-8        1.24ms ±10%  1.19ms ±11%  -3.46%  (p=0.000 n=46+42)
Filter/1000nodes/two-Constraints-zone-node/Clone-8  96.3µs ±21%  89.6µs ± 7%  -6.96%  (p=0.000 n=49+43
```

**Comment by yuzhiquan** [comment]
New benchmark result:
```shell
> /Users/yuzhiquan/go/bin/benchstat old_50-3.txt new_50-3.txt
name                                                old time/op  new time/op  delta
Filter/1000nodes/single-constraint-zone-8            886µs ±31%   751µs ±33%  -15.28%  (p=0.000 n=49+50)
Filter/1000nodes/single-constraint-zone/Clone-8     1.61µs ± 2%  1.54µs ± 3%   -4.25%  (p=0.000 n=47+42)
Filter/1000nodes/single-constraint-node-8            867µs ±19%   815µs ±31%   -6.06%  (p=0.000 n=46+50)
Filter/1000nodes/single-constraint-node/Clone-8     91.8µs ± 4%  88.3µs ± 2%   -3.77%  (p=0.000 n=45+44)
Filter/1000nodes/two-Constraints-zone-node-8        1.21ms ±12%  1.25ms ±34%     ~     (p=0.870 n=49+50)
Filter/1000nodes/two-Constraints-zone-node/Clone-8  90.1µs ± 2%  87.7µs ± 3%   -2.71%  (p=0.000 n=46+46)
```

**Comment by yuzhiquan** [comment]
Benchmark result:
```shell
> /Users/yuzhiquan/go/bin/benchstat old_50-4.txt new_50-4.txt
name                                                old time/op  new time/op  delta
Filter/1000nodes/single-constraint-zone-8            852µs ±47%   687µs ±18%  -19.39%  (p=0.000 n=50+47)
Filter/1000nodes/single-constraint-zone/Clone-8     1.64µs ±11%  1.56µs ± 4%   -5.15%  (p=0.000 n=44+45)
Filter/1000nodes/single-constraint-node-8            885µs ±40%   756µs ±18%  -14.55%  (p=0.000 n=44+49)
Filter/1000nodes/single-constraint-node/Clone-8     90.1µs ±11%  88.5µs ± 3%   -1.75%  (p=0.040 n=45+47)
Filter/1000nodes/two-Constraints-zone-node-8        1.15ms ±10%  1.15ms ±17%     ~     (p=0.511 n=49+47)
Filter/1000nodes/two-Constraints-zone-node/Clone-8  87.8µs ± 4%  87.0µs ± 2%   -0.92%  (p=0.000 n=43+47)
```

**Comment by alculquicondor** [comment]
A lot of tests are failing. Maybe you need to rebase?

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Pull request closed without merging**